### PR TITLE
Nisyad/add detection confusion matrix new

### PR DESCRIPTION
**Pull request by nisyad-ms**
Major refactor from previous related PR: https://github.com/microsoft/visionmetrics/pull/56

**Updated Definitions:**

 1. FP_due_to_wrong_class: IOU >= threshold but predicted class is different from the ground truth class.
 2. FP_due_to_low_iou: Predicted bbox IOU < threshold (including no overlap)
 3. fp_due_to_extra_pred_boxes: Excess predicted bboxes when all ground truth boxes have been matched

**Added test cases as suggested in PR comments.

**Pseudo code for counting TP, FP, FN:**
1. For each image, check empty conditions first. If preds or gt is empty, then assign to FN or FP. 
2. In an image, for each pred box, iterate over the GT boxes and find a TP. If TP is found (point 4) mark that GT box as used and not be used in susbsequent loops.
3. Keep track of best iou between the pred box and GT boxes, separately for same class and different class boxes.
4. If a same class gt box is found and best_same_class_iou >= thres, then mark as TP. Otherwise, if iou < thresh, mark as fp due to low iou
5. Otherwise, compare best_diff_class_iou with thresh and assign to fp due to wrong class or fp due to wrong iou
6. finally, count the remaining gt boxes for this image and assign them to fn
 

Note: conversation from previous PR copied below as a comment.

### CODE REVIEW

**Comment by nisyad-ms** [comment]
From @cy-bai previously, 
I think this still includes errors that are more like wrong class to me. For example, ground truth box [0,0,10,10] label 0, [90,90,100,100] label 1, prediction [90,90,100,100] label 0. In current prediction this belongs to "fp_due_to_wrong_class". How about for boxes that don't overlap w/ any gt, they belong to "no corresponding gt box", for boxes that have overlap but labels are wrong, they belong to "wrong class"?

File: visionmetrics/detection/confusion_matrix.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+            if self._is_empty(gts):
+                self.fp += len(preds)
+                self.fp_due_to_extra_pred_boxes += len(preds)
+                continue
+
+            # Count per image
+            gt_boxes_used = [False] * len(gts)  # To keep track of all matched GT boxes
+            preds = sorted(preds, key=lambda x: x[1], reverse=True)
+            for pred in preds:
```
Comment (gnloch) [inline visionmetrics/detection/confusion_matrix.py:80]: You are evaluating each prediction against each gt, which is kind of O(n²). I wonder if you could use something like this: https://msazure.visualstudio.com/VisionIntelligence/_git/RaaS.ModelTraining.Shared?path=/evaluation/detection_evaluation.py&version=GBmaster&line=395&lineEnd=396&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents to vectorize this op.

File: visionmetrics/detection/confusion_matrix.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
+import torch
+from torchmetrics import Metric
+
+
+class DetectionConfusionMatrix(Metric):
+    """
+        Calcuates confusion matrix in the context of object detection.
```
[the remark this answers is not in the document]
Comment (nisyad-ms) [reply visionmetrics/detection/confusion_matrix.py]: fixed.

File: visionmetrics/detection/confusion_matrix.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            TP: When a predicted bounding box has IoU greater than the threshold with a ground truth box of the same class
+            FN: When a ground truth bounding box has no corresponding predicted bounding box
+            FP:
+                1. FP_due_to_wrong_class: IOU >= threshold but predicted class is different from the ground truth class.
+                2. FP_due_to_low_iou: Predicted bbox IOU < threshold (including no overlap)
+                3. fp_due_to_extra_pred_boxes: Excess predicted bboxes when all ground truth boxes have been matched
+
+        """
+
+    def __init__(self, iou_threshold=0.5, **kwargs):
```
[the remark this answers is not in the document]
Comment (nisyad-ms) [reply visionmetrics/detection/confusion_matrix.py:37]: `predictions` and `targets` are added using the update() method in the Metric class.

File: visionmetrics/detection/confusion_matrix.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        self.add_state("fp", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fn", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp_due_to_wrong_class", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp_due_to_low_iou", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp_due_to_extra_pred_boxes", default=torch.tensor(0), dist_reduce_fx="sum")
+
+    def update(self, predictions, targets):
+        if len(predictions) != len(targets):
+            raise ValueError("Number of predictions and targets should be the same.")
+        if not isinstance(predictions[0][0], list):
```
[the remark this answers is not in the document]
Comment (nisyad-ms) [reply visionmetrics/detection/confusion_matrix.py]: good catch! updated.

File: visionmetrics/detection/confusion_matrix.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        super().__init__(**kwargs)
+        self.iou_threshold = iou_threshold
+        self.add_state("tp", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fn", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp_due_to_wrong_class", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp_due_to_low_iou", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp_due_to_extra_pred_boxes", default=torch.tensor(0), dist_reduce_fx="sum")
+
+    def update(self, predictions, targets):
```
[the remark this answers is not in the document]
Comment (nisyad-ms) [reply visionmetrics/detection/confusion_matrix.py]: added type hints

File: visionmetrics/detection/confusion_matrix.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+            'FP': self.fp.item(),
+            'FN': self.fn.item(),
+            'FP_due_to_wrong_class': self.fp_due_to_wrong_class.item(),
+            'FP_due_to_low_iou': self.fp_due_to_low_iou.item(),
+            'FP_due_to_extra_pred_boxes': self.fp_due_to_extra_pred_boxes.item()
+        }
+
+    def _update_confusion_matrix(self, predictions, targets, iou_threshold):
+        for preds, gts in zip(predictions, targets):
+
```
[the remark this answers is not in the document]
Comment (nisyad-ms) [reply visionmetrics/detection/confusion_matrix.py]: fixed

File: visionmetrics/detection/confusion_matrix.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+            # Count unused GT boxes as FNs
+            for gt_index, gt in enumerate(gts):
+                if not gt_boxes_used[gt_index]:
+                    self.fn += 1
+
+    @staticmethod
+    def _is_empty(pred_or_gt):
+        return all([len(x) == 0 for x in pred_or_gt])
+
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (nisyad-ms) [reply visionmetrics/detection/confusion_matrix.py:135]: torchmetrics*IOU excepts a list of dicts comprising tensors. Our preds and targets are nested lists of floats. So, I preferred to write a simple method to calculate IOU instead of doing float -> tensor conversions and back.

File: visionmetrics/detection/confusion_matrix.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+    def iou(box1, box2):
```
[the remark this answers is not in the document]
Comment (nisyad-ms) [reply visionmetrics/detection/confusion_matrix.py]: added

File: visionmetrics/detection/confusion_matrix.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+            # Count per image
+            gt_boxes_used = [False] * len(gts)  # To keep track of all matched GT boxes
+            preds = sorted(preds, key=lambda x: x[1], reverse=True)
+            for pred in preds:
+                # If all GT boxes for this image have been matched (TP)
+                # with corresponding pred boxes then all remaining pred boxes are FPs
+                if all(gt_boxes_used):
+                    self.fp += 1
+                    self.fp_due_to_extra_pred_boxes += 1
+                    continue
```
[the remark this answers is not in the document]
Comment (nisyad-ms) [reply visionmetrics/detection/confusion_matrix.py]: Yes, updated the logic.

File: visionmetrics/detection/confusion_matrix.py (hunk 1 above)
Comment (nisyad-ms) [reply visionmetrics/detection/confusion_matrix.py]: It's hard to vectorize since the inner loop is variable with multiple `continue` and `break` conditions. 
I can revisit if O(n2) becomes a bottleneck.

File: visionmetrics/detection/confusion_matrix.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+                    else:
+                        self.fp += 1
+                        self.fp_due_to_low_iou += 1
+                else:
+                    if best_diff_class_iou >= iou_threshold:
+                        self.fp += 1
+                        self.fp_due_to_wrong_class += 1
+                    else:
+                        self.fp += 1
+                        self.fp_due_to_low_iou += 1
```
Comment (cy-bai) [inline visionmetrics/detection/confusion_matrix.py:124]: For seek of detailed analysis, how about split this "fp_due_to_low_iou" into "fp_due_to_low_iou_and_wrong_class" (IOU>0),  and "fp_due_to_no_match"? (IOU==0)

Comment (nisyad-ms) [reply visionmetrics/detection/confusion_matrix.py:124]: **There are 8 cases to consider:**
1. iou >= thresh + correct class
2. iou >= thresh + wrong class
3. 0 < iou < thresh + correct class
4. 0 < iou < thresh + wrong class
5. iou = 0 + correct class
6. iou = 0 + wrong class
7. no gt box left to match
8. no pred box left to match

**Currently,**
Case 1 -> TP
Case 2 -> FP due to wrong class
Case 3,4,5,6 -> FP due to low IOU
Case 7 -> FP due to extra pred
Case 8 -> FN

I think if we split "fp_due_to_low_iou" into "fp_due_to_low_iou_and_wrong_class" and "fp_due_to_no_match", then we also need atleast "fp_due_to_low_iou_and_correct_class". The no_match (iou=0) case could also be split into correct and wrong classes. Basically, consider each case as a separate category? I suspect this

I think no match is just a subset of low_iou. So we could include no_match into low_iou cases depending on wrong and correct classes. 

TLDR: I suggest we split "fp_due_to_low_iou" into "fp_*_wrong_class" and "fp_*_correct_class" (these will also include the iou=0 cases)

Thoughts?

File: visionmetrics/detection/confusion_matrix.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gnloch) [reply visionmetrics/detection/confusion_matrix.py:37]: So `predictions` and `targets` are not part of the `kwargs` here? So why are they on the `__init__` docstring?

[the remark this answers is not in the document]
Comment (nisyad-ms) [reply visionmetrics/detection/confusion_matrix.py:33]: I just intended to provide an overall info on what the class does. I agree this could be confusing for first time user. 
I rephrased the class docstring for more clarity

File: visionmetrics/detection/confusion_matrix.py (hunk 10 above)
Comment (nisyad-ms) [reply visionmetrics/detection/confusion_matrix.py]: @cy-bai - thoughts?

**Review by cy-bai — APPROVED** [review]

File: visionmetrics/detection/confusion_matrix.py (hunk 10 above)
Comment (cy-bai) [reply visionmetrics/detection/confusion_matrix.py]: LGTM thanks!
